common/src/main/java/io/druid/collections/StupidPool.java

```diff
@@ -99,29 +136,45 @@ public T get()
     @Override
     public void close()
     {
-      if (!closed.compareAndSet(false, true)) {
-        log.warn(new ISE("Already Closed!"), "Already closed");
-        return;
-      }
-      if (objects.size() < objectsCacheMaxCount) {
-        if (!objects.offer(object)) {
-          log.warn(new ISE("Queue offer failed"), "Could not offer object [%s] back into the queue", object);
-        }
+      final T object = objectRef.get();
+      if (object != null && objectRef.compareAndSet(object, null)) {
+        tryReturnToPool(object);
+        // Effectively does nothing, because objectRef is already set to null. The purpose of this call is to
+        // deregister the cleaner from the internal linked list of all cleaners in the JVM.
+        cleaner.clean();
       } else {
-        log.debug("cache num entries is exceeding max limit [%s]", objectsCacheMaxCount);
+        log.warn(new ISE("Already Closed!"), "Already closed");
       }
     }
+  }
+
+  private class ObjectReclaimer implements Runnable
+  {
+    private final AtomicReference<T> objectRef;
+
+    private ObjectReclaimer(AtomicReference<T> objectRef)
+    {
+      this.objectRef = objectRef;
+    }
 
     @Override
-    protected void finalize() throws Throwable
+    public void run()
     {
       try {
-        if (!closed.get()) {
+        final T object = objectRef.get();
+        if (object != null && objectRef.compareAndSet(object, null)) {
           log.warn("Not closed!  Object was[%s]. Allowing gc to prevent leak.", object);
+          tryReturnToPool(object);
```

Comment:
earlier we did not use to return the object back to pool if the holder wasn't closed. and, it is probably safe to not reuse it given that it wasn't explicitly returned back.